File: web/src/solver/gaussSeidel.ts
import type { FieldConfig, SolveResult } from '../types/simulation';

/**
 * TypeScript reference Gauss-Seidel solver.
 * Runs synchronously — used for Phase 1 before WASM worker is ready.
 */
export function solveGaussSeidel(config: FieldConfig): SolveResult {
  const { width, height } = config.grid;
  const potential = new Float32Array(width * height);
  const rhs = new Float32Array(width * height);

  // Set charge density
  for (const charge of config.charges) {
    const ix = Math.round(charge.x);
    const iy = Math.round(charge.y);
    if (ix >= 0 && ix < width && iy >= 0 && iy < height) {
      rhs[iy * width + ix] = -charge.q;
    }
  }

  // Apply initial boundary conditions
  applyBoundary(potential, config.boundary, width, height);

  const t0 = performance.now();
  let iterations = 0;
  let residual = Infinity;
  let converged = false;

  for (let iter = 0; iter < config.maxIterations; iter++) {
    residual = 0;

    for (let y = 1; y < height - 1; y++) {
      for (let x = 1; x < width - 1; x++) {
        const idx = y * width + x;
        const old = potential[idx];
        const newVal = 0.25 * (
          potential[idx + 1] +
          potential[idx - 1] +
          potential[(y + 1) * width + x] +
          potential[(y - 1) * width + x] -
          rhs[idx]
        );
        const diff = Math.abs(newVal - old);
        if (diff > residual) residual = diff;
        potential[idx] = newVal;
      }
    }

    applyBoundary(potential, config.boundary, width, height);
    iterations++;

    if (residual < config.tolerance) {
      converged = true;
      break;
    }
  }

  return {
    potential,
    width,
    height,
    iterations,
    residual,
    converged,
    timeMs: performance.now() - t0,
  };
}

function applyBoundary(
  grid: Float32Array,
  bc: FieldConfig['boundary'],
  w: number,
  h: number,
): void {
  switch (bc.type) {
    case 'dirichlet_zero':
      for (let x = 0; x < w; x++) {
        grid[x] = 0;
        grid[(h - 1) * w + x] = 0;
      }
      for (let y = 0; y < h; y++) {
        grid[y * w] = 0;
        grid[y * w + w - 1] = 0;
      }
      break;
    case 'dirichlet_fixed': {
      const [top, right, bottom, left] = bc.voltages;
      for (let x = 0; x < w; x++) {
        grid[x] = top;
        grid[(h - 1) * w + x] = bottom;
      }
      for (let y = 0; y < h; y++) {
        grid[y * w] = left;
        grid[y * w + w - 1] = right;
      }
      break;
    }
    case 'neumann':
      for (let x = 1; x < w - 1; x++) {
        grid[x] = grid[w + x];
        grid[(h - 1) * w + x] = grid[(h - 2) * w + x];
      }
      for (let y = 1; y < h - 1; y++) {
        grid[y * w] = grid[y * w + 1];
        grid[y * w + w - 1] = grid[y * w + w - 2];
      }
      break;
  }
}

File: solver/src/field/gauss_seidel.rs
use crate::common::config::{BoundaryCondition, FieldConfig};
use crate::common::grid::Grid;
use super::engine::{FieldSolver, StepResult};

pub struct GaussSeidelSolver {
    potential: Grid,
    rhs: Grid,
    config: FieldConfig,
    total_iterations: usize,
}

impl GaussSeidelSolver {
    pub fn new(config: &FieldConfig) -> Self {
        let w = config.width;
        let h = config.height;
        let mut potential = Grid::new(w, h);
        let mut rhs = Grid::new(w, h);

        for charge in &config.charges {
            if charge.x < w && charge.y < h {
                rhs.set(charge.x, charge.y, -charge.q);
            }
        }

        apply_boundary(&mut potential, &config.boundary, w, h);

        Self {
            potential,
            rhs,
            config: config.clone(),
            total_iterations: 0,
        }
    }
}

impl FieldSolver for GaussSeidelSolver {
    fn step(&mut self, n: usize) -> StepResult {
        let w = self.config.width;
        let h = self.config.height;
        let mut residual = 0.0;

        for _ in 0..n {
            residual = 0.0;

            for y in 1..h - 1 {
                for x in 1..w - 1 {
                    let old = self.potential.get(x, y);
                    let new_val = 0.25 * (
                        self.potential.get(x + 1, y)
                        + self.potential.get(x - 1, y)
                        + self.potential.get(x, y + 1)
                        + self.potential.get(x, y - 1)
                        - self.rhs.get(x, y)
                    );
                    let diff = (new_val - old).abs();
                    if diff > residual {
                        residual = diff;
                    }
                    self.potential.set(x, y, new_val);
                }
            }

            apply_boundary(&mut self.potential, &self.config.boundary, w, h);
            self.total_iterations += 1;
        }

        StepResult {
            residual,
            converged: residual < self.config.tolerance,
        }
    }

    fn potential(&self) -> &Grid {
        &self.potential
    }

    fn iterations(&self) -> usize {
        self.total_iterations
    }
}

fn apply_boundary(grid: &mut Grid, bc: &BoundaryCondition, w: usize, h: usize) {
    match bc {
        BoundaryCondition::DirichletZero => {
            for x in 0..w {
                grid.set(x, 0, 0.0);
                grid.set(x, h - 1, 0.0);
            }
            for y in 0..h {
                grid.set(0, y, 0.0);
                grid.set(w - 1, y, 0.0);
            }
        }
        BoundaryCondition::DirichletFixed { voltages } => {
            let [top, right, bottom, left] = *voltages;
            for x in 0..w {
                grid.set(x, 0, top);
                grid.set(x, h - 1, bottom);
            }
            for y in 0..h {
                grid.set(0, y, left);
                grid.set(w - 1, y, right);
            }
        }
        BoundaryCondition::Neumann => {
            for x in 1..w - 1 {
                let v = grid.get(x, 1);
                grid.set(x, 0, v);
                let v = grid.get(x, h - 2);
                grid.set(x, h - 1, v);
            }
            for y in 1..h - 1 {
                let v = grid.get(1, y);
                grid.set(0, y, v);
                let v = grid.get(w - 2, y);
                grid.set(w - 1, y, v);
            }
        }
    }
}
